[DROOLS-4102] Persist/reload Test Scenario table width/height

yesamer opened the pull request:
@kkufova @gitgabrio @danielezonca can you please test and review it?

With this PR, it's now possible to persist / load columns width.
In .scesim file, this parameter is saved as child of tag <factMapping>.
Please test, in particular the following cases (both DMN / RULE scenarios)

-     Create/save a simulation (default column width/row height in scesim)
-     Rename a simulation (it perform a save/reload)
-     Change column type (recreate the column)
-     Duplicate column (preserve same size)
-     Insert/Delete a column
-     History (undo and redo actions)

Thanks.

Related PR: https://github.com/kiegroup/drools/pull/2476

https://issues.jboss.org/browse/DROOLS-4102
https://issues.jboss.org/browse/DROOLS-4083

In draft because Unit Tests are missing, but you can start to review / test it.

yesamer commented on drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/handlers/ScenarioGridWidgetDnDMouseMoveHandler.java:
```diff
@@ -0,0 +1,52 @@
+/**
+ * MouseMoveHandler to handle column resize event.
+ */
+public class ScenarioGridWidgetDnDMouseMoveHandler extends GridWidgetDnDMouseMoveHandler {
```
This could be renamed as ScenarioGridWidgetColumnResize, as alternative.

gitgabrio replied:
@yesamer 
if the parent' name is GridWidgetDnDMouseMoveHandler, it makes sense to add a "specification" to it, and not give a name with a different meaning. That class is actually a Drag'n'drop handler

yesamer commented on drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/models/ScenarioGridModel.java:
```diff
@@ -243,7 +244,11 @@ public void duplicateColumnValues(int originalColumnIndex, int newColumnIndex) {
+        column.setIndex(columns.size());
+        columns.add(index, column);
+
+        selectionsManager.onInsertColumn(index);
```
This method (and in particular the last 3 lines) could be put in its superclass.

yesamer commented:
@kkufova Ready to be tested. Please, contact me in case of futher information.

kkufova commented:
We agreed with @yesamer on the following:

- The width of newly added columns should be fixed (= the minimal column width).
- The column width should **not** be affected in any way by setting a column mapping or undo/redo operations. I found a bug in this behavior; see [this screencast](https://issues.jboss.org/secure/attachment/12453820/screencast-01-08-19-1.webm) for details.

kkufova commented:
@yesamer, I checked the newest version and the width persistence doesn't work anymore; changed width is lost after saving and undo.

kkufova commented:
@yesamer, is this ready for another review round?

yesamer commented:
@kkufova ok, i need only to finish unit tests part.

yesamer commented:
@kkufova you can proceed :)

kkufova commented:
@yesamer, deleting columns still causes inconsistencies with the previous state.

1. The column's width is completely changed after deleting the only EXPECT column (a screencast can be found [here](https://issues.jboss.org/secure/attachment/12454033/screencast-07-08-19-1.webm)). We can discuss this further, but I think that deleting a column should not change the column width (if you are only deleting the mapping in case of a single available column).
2. Adding and deleting a column slightly changes the width of existing columns (a screencast can be found [here](https://issues.jboss.org/secure/attachment/12454034/screencast-07-08-19-2.webm)). I think this should not happen.

I feel like this is getting more and more complicated. Let me know what you think.

yesamer commented:
@kkufova 
FIXED --> The column's width is completely changed after deleting the only EXPECT column (a screencast can be found here). We can discuss this further, but I think that deleting a column should not change the column width (if you are only deleting the mapping in case of a single available column).

As agreed, i reverted the last changes I added for undo/redo handling.

gitgabrio commented on drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/models/ScenarioGridModel.java:
```diff
@@ -266,6 +267,7 @@ public void deleteColumn(int columnIndex) {
         final GridColumn<?> toDelete = getColumns().get(columnIndex);
         deleteColumn(toDelete);
         simulation.removeFactMappingByIndex(columnIndex);
+        synchronizeFactMappingsWidths();
```
@yesamer 
I think that's to be removed. `synchronizeFactMappingsWidths` put the current columns widht inside factmappings, but we decided it should happen **only** when the user saves the scenario - otherwise rely on the "status" implementation to manage them

yesamer replied:
Done!

gitgabrio commented on drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/models/ScenarioGridModel.java:
```diff
@@ -318,8 +320,9 @@ protected void replaceColumn(int columnIndex, GridColumn<?> column) {
         String columnId = ((ScenarioGridColumn) column).getInformationHeaderMetaData().getColumnId();
         ExpressionIdentifier ei = ExpressionIdentifier.create(columnId, FactMappingType.valueOf(group));
         commonAddColumn(columnIndex, column, ei);
+        /* Restoring the expected columns dimension, overriding the automatic resizing */
```
@yesamer 
I personally do not have nothing against such comments, but keep in mind not everyone agrees on that.

yesamer replied:
I'll keep it in mind.

gitgabrio commented on drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/models/ScenarioGridModel.java:
```diff
@@ -515,6 +518,58 @@ public void clearSelections() {
+    /**
+     * It forces a <code>internalRefreshWidth</code> refresh
+     */
+    public boolean forceRefreshWidth() {
```
@yesamer 
Please check it is really needed - columns width are already automatically managed, and persisting them on our model should happen only at "save" invocation

yesamer replied:
@gitgabrio In my opinion this is still required, because it's necessary to launch AutoResize process after loading of column widths from scesim file.

gitgabrio commented on drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/models/ScenarioGridModel.java:
```diff
@@ -515,6 +518,58 @@ public void clearSelections() {
+     * @param column
+     * @param width
+     */
+    public void updateColumnWidth(final GridColumn<?> column, final double width) {
```
@yesamer 
Same comment as above

yesamer replied:
Done.

gitgabrio commented on drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/models/ScenarioGridModel.java:
```diff
@@ -515,6 +518,58 @@ public void clearSelections() {
+            final FactMapping factMapping = simulation.getSimulationDescriptor().getFactMappingByIndex(columnIndex);
+            if (factMapping.getColumnWidth() != null) {
+                column.setWidth(factMapping.getColumnWidth());
+            } else {
```
@yesamer 
I think this `else` branch is not needed (see above comments)

yesamer replied:
Right, done.

gitgabrio commented on drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/widgets/ScenarioGrid.java:
```diff
@@ -73,6 +73,8 @@ public void setContent(Simulation simulation) {
         ((ScenarioGridModel) model).bindContent(simulation);
         setHeaderColumns(simulation);
         appendRows(simulation);
+        ((ScenarioGridModel) model).loadFactMappingsWidth();
```
@yesamer 
Please remove those two lines and set the column' width directly at column creation inside `ScenarioGrid.setHeaderColumn` (I think it should be possible))

yesamer replied:
@gitgabrio As explained, this is required because, during the insertion of the column the grid, the current logic re-calculate the column dimension based on the number of the columns present in the grid.

yesamer commented:
@gitgabrio @danielezonca Fine for me to be merged.
